packages/components/src/custom-select-control-v2/index.tsx

```diff
@@ -0,0 +1,72 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import * as Ariakit from '@ariakit/react';
+/**
+ * WordPress dependencies
+ */
+import { createContext, useContext, useMemo } from '@wordpress/element';
+/**
+ * Internal dependencies
+ */
+import { useCx } from '../utils/hooks/use-cx';
+import * as Styled from './styles';
+import type {
+	CustomSelectProps,
+	CustomSelectItemProps,
+	CustomSelectContext as CustomSelectContextType,
+} from './types';
+
+export const CustomSelectContext =
+	createContext< CustomSelectContextType >( undefined );
+
+export function CustomSelect( props: CustomSelectProps ) {
+	const {
+		children,
+		defaultValue,
+		label,
+		onChange,
+		size = 'default',
+		value,
+		styledValue,
+	} = props;
+
+	const store = Ariakit.useSelectStore( {
+		setValue: ( nextValue ) => onChange?.( nextValue ),
+		defaultValue,
+		value,
+	} );
+
+	const { value: currentValue } = store.useState();
+
+	const cx = useCx();
+
+	const classes = useMemo(
+		() => cx( Styled.inputSize[ size ] ),
+		[ cx, size ]
+	);
+
+	return (
+		<CustomSelectContext.Provider value={ { store } }>
+			<Ariakit.SelectLabel store={ store }>{ label }</Ariakit.SelectLabel>
+			<Styled.CustomSelectButton className={ classes } store={ store }>
+				{ styledValue ? styledValue( currentValue ) : currentValue }
```

Comment:
We should probably add some default fallback text for when no item is selected — I know that Ariakit does its best to always pick an item, but I wonder what would happen if the `value` prop passed to the component was `null` (or even a string that doesn't match any existing item)

```suggestion
				{ styledValue ? styledValue( currentValue ) : (currentValue ?? "Select an item") }
```